pkg/broker/deprovision_subscriber.go

```diff
@@ -36,64 +35,40 @@ func NewDeprovisionWorkSubscriber(dao *dao.Dao) *DeprovisionWorkSubscriber {
 // Subscribe - will start the work subscriber listenning on the message buffer for deprovision messages.
 func (d *DeprovisionWorkSubscriber) Subscribe(msgBuffer <-chan JobMsg) {
 	d.msgBuffer = msgBuffer
-
 	go func() {
 		log.Info("Listening for deprovision messages")
-		for {
-			msg := <-msgBuffer
-			metrics.DeprovisionJobFinished()
-			log.Debug("Processed deprovision message from buffer")
-
-			if msg.Error != "" {
-				// Job failed, mark failure
-				log.Errorf("Deprovision job reporting error: %s", msg.Error)
-				setFailedDeprovisionJob(d.dao, msg)
-				continue
-			}
-
-			instance, err := d.dao.GetServiceInstance(msg.InstanceUUID)
-			if err != nil {
-				log.Errorf(
-					"Error occurred getting service instance [ %s ] after deprovision job:",
-					msg.InstanceUUID,
-				)
-				setFailedDeprovisionJob(d.dao, msg)
-				continue
+		for msg := range msgBuffer {
+			log.Debug("received deprovision message from buffer")
+			if msg.State.State == apb.StateSucceeded {
```

Comment:
I would flip the check to look for the error conditions first and have escape valves, then do the success case afterwards without being in an if block.

```golang
func (d *DeprovisionWorkSubscriber) Subscribe(msgBuffer <-chan JobMsg) {
    d.msgBuffer = msgBuffer
    go func() {
        log.Info("Listening for deprovision messages")
        for msg := range msgBuffer {
            log.Debug("received deprovision message from buffer")

            if msg.State.State != apb.StateSucceeded { 
                if err := d.dao.SetState(msg.InstanceUUID, msg.State); err != nil {
                    log.Errorf("failed to set state after deprovision %v", err)
                }
                continue
            }

            instance, err := d.dao.GetServiceInstance(msg.InstanceUUID)
            if err != nil {
                log.Errorf(
                    "Error occurred getting service instance [ %s ] after deprovision job:",
                    msg.InstanceUUID,
                )   
                setFailedDeprovisionJob(d.dao, msg)
                continue
            }
            if err := cleanupDeprovision(instance, d.dao); err != nil {
                log.Errorf("Failed cleaning up deprovision after job, error: %v", err)
                // Cleanup is reporting something has gone wrong. Deprovision overall
                // has not completed. Mark the job as failed.
                setFailedDeprovisionJob(d.dao, msg)
                continue
            }   
        }
    }()
}

```

Reply:
👍

Reply:
removed the if entirely as we always want the state to be set